### ISSUE TITLE
Skin and bodygroup options for player_manager.AddValidModel

### ISSUE THREAD

**Issue by TiberiumFusion** [Facepunch/garrysmod-requests#1801]
## Details
`AddValidHands()` can already do this. `AddValidModel()` cannot. Given how Gmod 13 is about 9 years old now and this feature _still_ does not yet exist, I wouldn't be surprised at all if this is due the engine binding a player's chosen playermodel to merely the mdl path instead of some component abstraction layer. Needless to say, that is a very poor design choice and could stand be changed.

### Viewmodel hands and `AddValidHands()`
• Let's say I have a viewmodel hands model that's 3MB in size. It has 12 possible bodygroup combinations and 4 skins.
• `AddValidHands()` lets me pick the skin and bodygroup for each playermodel that the viewmodel hands are assigned to.
• I only have include that one model in my addon and it covers all 48 bodygroup+skin combos.
◆ Final cost: 3MB
◆ That's 0MB of wasted space or 100% efficiency.

### Playermodels and `AddValidModel()`
• Let's say I have a playermodel that's 7MB in size. It has 12 possible bodygroup combinations and 4 skins. For the sake of demonstration, all bodygroup combinations affect the model's arms and thus must be reflected in the viewmodel hands.
• `AddValidModel()` does not accept any initial skin or initial bodygroup configuration.
• I have to recompile the model 48 times with different material names and bodygroup mesh orders to cover all 48 bodygroup+skin combos.
◆ Final cost: 7 * 48 = 336MB
◆ That's 329MB of wasted space in my addon or 2% efficiency. Not only does this waste an inordinate amount of space on the player's drive, but the addon will load up to as 48x slower as well. Inevitable fragmentation of these absurdly large (and common) gma files, as well as with the messy addons cache directory itself, can easily triple loading times on top of that.

### Bandaid addons
These are addons which:
• Allow the player to set the bodygroups and skins of their playermodel (i.e. Enhanced Playermodel Selector)
• Automatically set the playermodel bodygroups and skins in accordance with some rules (e.g. DarkRP jobs and their specific playermodels)
I call these **bandaid** addons because that is what they are. They require more, unnecessary work from either the gamemode author and/or the player, require ubiquitous _awareness of their existence_ in the first place (doesn't exist), and are liable to manipulation and case-dependent cheating. Additionally, they do not handle viewmodel hands, and so any player who adjusts their playermodel's skins and bodygroups will now have incorrect viewmodel hands.

### Lose-lose situation
The bottom line is that bodygroups and skins are nearly worthless in gmod's playermodels. It becomes less costly to just compile 48 semi-unique models with 0 bodygroups and 0 alternate skins. But, in doing so, that screws over the ability to use the models as ragdolls/NPCs efficiently. Now you have to delete and spawn a new model just to get a different bodygroup and skin combo, and then deal with the headaches of reassociating the new model with everything else the old one was associated with.

**Comment by robotboy655** [comment]
Uh, with player models you choose the skins and bodygroups yourself tho.

**Comment by thegrb93** [comment]
Related https://github.com/Facepunch/garrysmod/pull/1525

**Comment by thegrb93** [comment]
@robotboy655

**Comment by TiberiumFusion** [comment]
@robotboy655 You have entirely missed the point.
I don't care if the user wants to change their bodygroups and skins after picking a playermodel. The goal is to ship a single model that utilizes bodygroups and skins to represent many different "logical" playermodels. The game should recognize each logical playermodel as indepedent choices, just like any other playermodel, and present them all to the player in the playermodel selector as indepdedent choices, just like any other playermodel.

Not every server has the Enhanced Playermodel Selector. Not every server allows playermodel customization. Not every player even knows that they can set bodygroups and skins with addons like Enhanced Playermodel Selector. Not every gamemode allows playermodel customization. Not every gamemode or server makes it accessible. And not every server expects their players to fiddle around with their bodygroups and skins to have access to all available appearances, conform to the rules, etc.

Creating a different playermodel for all (or most) different bodygroup + skin combos is the most direct solution to every aforementioned issue. Player opens playermodel selector -> player sees icon they want -> player clicks icon -> player gets playermodel. Easy done, no problems. But this comes at the cost of extremely inflated addon size. Changing gmod from using a simple MDL path to using a proper playermodel description - with an initial skin and bodygroup selection - immediately fixes this issue.
